quality-debt: .agents/scripts/screen-time-helper.sh — PR #3997 review feedback (medium)

## Unactioned Review Feedback

**Source PR**: #3997
**File**: `.agents/scripts/screen-time-helper.sh`
**Reviewers**: gemini
**Findings**: 1
**Max severity**: medium

---

### MEDIUM: gemini (gemini-code-assist[bot])
**File**: `.agents/scripts/screen-time-helper.sh:714`
![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)

To improve performance and script efficiency, it's best to consolidate the multiple `jq` calls that read from `$HISTORY_FILE` into a single pass. You can compute all the necessary historical statistics (1-day, 7-day, 28-day, 365-day, and total) in one `jq` invocation and then read them into shell variables. This avoids reading and parsing the same file multiple times, which is especially beneficial if the history file grows large.

```shell
 local history_days=0 history_total=0
 local hist_1d=0 hist_7d=0 hist_28d=0 hist_365d=0
 if [[ -f "$HISTORY_FILE" ]]; then
  history_days=$(wc -l <"$HISTORY_FILE" | tr -d ' ')
  if [[ "$history_days" -gt 0 ]]; then
   # Read all history stats in one jq call for efficiency
   local history_stats
   history_stats=$(jq -s '
    . as $all |
    [
     ([$all[].screen_hours] | add // 0),
     ([$all[-1:][].screen_hours] | add // 0 | . * 10 | round / 10),
     ([$all[-7:][].screen_hours] | add // 0 | . * 10 | round / 10),
     ([$all[-28:][].screen_hours] | add // 0 | . * 10 | round / 10),
     ([$all[-365:][].screen_hours] | add // 0 | . * 10 | round / 10)
    ] | @tsv
   ' "$HISTORY_FILE" 2>/dev/null)

   # shellcheck disable=SC2086
   read -r history_total hist_1d hist_7d hist_28d hist_365d <<<"$history_stats"
  fi
 fi

 # If live 7d/28d returned 0 but we have history, use history instead
 if [[ "$history_days" -gt 0 ]]; then
  if [[ "$(echo "$week_hours == 0" | bc)" -eq 1 ]]; then
   week_hours="${hist_7d:-0}"
  fi
  if [[ "$(echo "$month_hours == 0" | bc)" -eq 1 ]]; then
   month_hours="${hist_28d:-0}"
  fi
  if [[ "$(echo "$day_hours == 0" | bc)" -eq 1 ]]; then
   day_hours="${hist_1d:-0}"
  fi
 fi

 # For 365 days: use accumulated history if available, else extrapolate
 local year_hours
 if [[ "$history_days" -gt 0 ]]; then
  local daily_avg
  daily_avg=$(echo "scale=2; ${history_total:-0} / $history_days" | bc)
  if [[ "$history_days" -ge 365 ]]; then
   # Have a full year of data — use last 365 days from history
   year_hours="${hist_365d:-0}"
  else
   # Extrapolate from available data
   year_hours=$(echo "scale=1; $daily_avg * 365" | bc)
  fi
```

<details>
<summary>References</summary>

1. Consolidate multiple 'jq' calls into a single pass where possible to improve performance and script efficiency.
</details>

[View comment](https://github.com/marcusquinn/aidevops/pull/3997#discussion_r2907093028)
---

---
_Auto-generated by `quality-feedback-helper.sh scan-merged`. Review each finding and either fix the code or dismiss with a reason._

--- a/.agents/scripts/screen-time-helper.sh
+++ b/.agents/scripts/screen-time-helper.sh
@@ -644,12 +644,12 @@ cmd_history() {
 		history_stats=$(jq -rs '
 			[
 				([.[].screen_hours] | add // 0 | . * 10 | round / 10),
-				(min_by(.date) | .date),
-				(max_by(.date) | .date)
+				(min_by(.date) | .date // "unknown"),
+				(max_by(.date) | .date // "unknown")
 			] | @tsv
 		' "$HISTORY_FILE" 2>/dev/null) || true
 		if [[ -n "$history_stats" ]]; then
-			read -r total_hours earliest latest <<<"$history_stats"
+			IFS=$'\t' read -r total_hours earliest latest <<<"$history_stats"
 		fi
 	fi
 

--- a/.agents/scripts/screen-time-helper.sh
+++ b/.agents/scripts/screen-time-helper.sh
@@ -633,13 +633,25 @@ cmd_history() {
 	fi
 
 	local total_days
-	local total_hours
-	local earliest
-	local latest
 	total_days=$(wc -l <"$HISTORY_FILE" | tr -d ' ')
-	total_hours=$(jq -s '[.[].screen_hours] | add | . * 10 | round / 10' "$HISTORY_FILE" 2>/dev/null || echo "0")
-	earliest=$(jq -s 'min_by(.date) | .date' "$HISTORY_FILE" 2>/dev/null || echo "unknown")
-	latest=$(jq -s 'max_by(.date) | .date' "$HISTORY_FILE" 2>/dev/null || echo "unknown")
+
+	# Single jq pass for all history stats (total_hours, earliest, latest)
+	local total_hours="0"
+	local earliest="unknown"
+	local latest="unknown"
+	if [[ "$total_days" -gt 0 ]]; then
+		local history_stats
+		history_stats=$(jq -rs '
+			[
+				([.[].screen_hours] | add // 0 | . * 10 | round / 10),
+				(min_by(.date) | .date),
+				(max_by(.date) | .date)
+			] | @tsv
+		' "$HISTORY_FILE" 2>/dev/null) || true
+		if [[ -n "$history_stats" ]]; then
+			read -r total_hours earliest latest <<<"$history_stats"
+		fi
+	fi
 
 	echo "Screen time history: ${total_days} days, ${total_hours}h total"
 	echo "Range: ${earliest} to ${latest}"

